File: LeetCode/38. 外观数列.js
/**
 * @param {number} n
 * @return {string}
 */
var countAndSay = function (n) {
  let str = '1'
  for (let i = 2; i <= n; i++) {
    let position = 0
    let start = 0
    const db = []
    while (position < str.length) {
      while (str[position] === str[start]) {
        position++
      }
      db.push('' + (position - start) + str[start])
      start = position
    }
    str = db.join('')
  }
  return str
}

File: LeetCode/1-100/38. 外观数列.java
class Solution {
    public String countAndSay(int n) {
    String res = "1";
        for (int i = 2; i <= n; i++) {
            int position = 0,start = 0;
            StringBuilder sb = new StringBuilder();
            int sl = res.length();
            while (position < sl) {
                while (res.charAt(start) == res.charAt(position)) {
                    position++;
                    if (position == sl) break;
                }
                String s = (position - start) + "" + res.charAt(start);
                sb.append(s);
                start = position;
            }
            res = sb.toString();
        }
        return res;
    }
}